### PR TITLE
test: use playwright for integration tests

### PR DESCRIPTION
**Pull request by mcansh**
the main differences with playwright compared to jest + puppeteer are
- the lack of a "watch" mode
- no `toMatchInlineSnapshot`
- describe, beforeTest, beforeEach, etc need to be `test.` prefixed
- all tests need to use `test` and not `it`

Signed-off-by: Logan McAnsh <logan@mcan.sh>

Closes: #

- [ ] Docs
- [ ] Tests

### CODE REVIEW

**Comment by mcansh** [comment]
gonna look into setting up a playwright "fixture" so the api is a bit closer to what we have with puppeteer.

so instead of passing `page` from playwright to all the `app.whatever` calls like this:
```ts
test("handles files under upload size limit", async ({ page }) => {
    let uploadFile = path.join(
      fixture.projectDir,
      "toUpload",
      "underLimit.txt"
    );
    let uploadData = Array(1_000).fill("a").join(""); // 1kb
    await fs
      .mkdir(path.dirname(uploadFile), { recursive: true })
      .catch(() => {});
    await fs.writeFile(uploadFile, uploadData, "utf8");

    await app.goto(page, "/file-upload");
    await app.uploadFile(page, "#file", uploadFile);
    await app.clickSubmitButton(page, "/file-upload");
})
```

we can instead do this

```ts
test("handles files under upload size limit", async ({ page, idkthenameyet }) => {
    let uploadFile = path.join(
      fixture.projectDir,
      "toUpload",
      "underLimit.txt"
    );
    let uploadData = Array(1_000).fill("a").join(""); // 1kb
    await fs
      .mkdir(path.dirname(uploadFile), { recursive: true })
      .catch(() => {});
    await fs.writeFile(uploadFile, uploadData, "utf8");

    await idkthenameyet.goto("/file-upload");
    await idkthenameyet.uploadFile("#file", uploadFile);
    await idkthenameyet.clickSubmitButton("/file-upload");
})
```

File: integration/helpers/playwright-fixture.ts [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+   */
+  collectResponses(filter?: UrlFilter) {
+    return collectResponses(this.page, filter);
+  }
+
+  /**
+   * Get HTML from the page. Useful for asserting something rendered that
+   * you expected.
+   *
+   * @param page The page from the Playwright context
```
Comment (F3n67u) [inline integration/helpers/playwright-fixture.ts:159]:
```suggestion
```

File: integration/helpers/playwright-fixture.ts [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+   * @param selector CSS Selector for the element's HTML you want
+   */
+  getHtml(selector?: string) {
+    return getHtml(this.page, selector);
+  }
+
+  /**
+   * Get a cheerio instance of an element from the page.
+   *
+   * @param page The page from the Playwright context
```
Comment (F3n67u) [inline integration/helpers/playwright-fixture.ts:169]:
```suggestion
```

File: integration/helpers/playwright-fixture.ts [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+   * @param selector CSS Selector for the element's HTML you want
+   */
+  async getElement(source?: string, selector?: string) {
```
Comment (F3n67u) [inline integration/helpers/playwright-fixture.ts:172]:
```suggestion
  async getElement(selector: string) {
```

Comment (F3n67u) [reply integration/helpers/playwright-fixture.ts:172]: this change will make the integration tests all green.

File: packages/remix-dev/package.json [hunk 4]
```diff
@@ -29,6 +29,7 @@
     "esbuild": "0.14.22",
     "esbuild-plugin-cache": "^0.2.9",
     "execa": "^5.1.1",
+    "esbuild-register": "^3.3.2",
```
Comment (kentcdodds) [inline packages/remix-dev/package.json:32]: Is this needed as a dependency?

Comment (mcansh) [reply packages/remix-dev/package.json:32]: i think it's from a merge conflict, i can remove it

Comment (kentcdodds) [reply packages/remix-dev/package.json:32]: On that note, how much trouble would it be to clean up the commit history on this branch? If it takes more than 10 seconds, don't bother.

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -107,5 +107,15 @@ jobs:
         # because we have a yarn.lock and caches efficiently.
         uses: bahmutov/npm-install@v1
 
+      - name: 📥 Install Playwright
+        run: npx playwright install --with-deps
```
Comment (MichaelDeBoey) [inline .github/workflows/test.yml:111]: Shouldn't we add this as a dependency to the repo, so we can run integration tests locally? 🤔

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -107,5 +107,15 @@ jobs:
...
+
       - name: 👀 Run Integration Tests
         run: "yarn test:integration"
+
+      - name: 📤 Upload test results
+        if: always()
+        uses: actions/upload-artifact@v2
```
Comment (MichaelDeBoey) [inline .github/workflows/test.yml:118]: v3 is already released

```suggestion
        uses: actions/upload-artifact@v3
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml (hunk 5 above)
Comment (mcansh) [reply .github/workflows/test.yml:111]: not according to their docs, this way we get the proper browser for linux https://playwright.dev/docs/ci#github-actions

File: .github/workflows/test.yml (hunk 6 above)
Comment (mcansh) [reply .github/workflows/test.yml:118]: just following the playwright docs 😅

File: packages/remix-dev/package.json (hunk 4 above)
Comment (mcansh) [reply packages/remix-dev/package.json:32]: should be pretty easy to do, i'll check after lunch

File: .github/workflows/test.yml (hunk 6 above)
Comment (MichaelDeBoey) [reply .github/workflows/test.yml]: Just wanted to let you know 😉

File: .gitignore [hunk 7]
```diff
@@ -9,6 +9,8 @@ yarn-error.log
 /fixtures/test
 /fixtures/my-remix-app
 /fixtures/deno-app
+/playwright-report
+/test-results
```
Comment (kentcdodds) [inline .gitignore:13]: When does this folder show up? I'm not getting it locally.

File: integration/action-test.ts [hunk 8]
```diff
@@ -254,56 +262,54 @@ describe("actions", () => {
   // TODO: figure out what the heck is wrong with this test...
   // For some reason the error message is "Unexpect Server Error" in the test
   // but if you try the app in the browser it works as expected.
-  it.skip("rejects too big of an upload with JavaScript", async () => {
+  test.skip("rejects too big of an upload with JavaScript", async ({
```
Comment (kentcdodds) [inline integration/action-test.ts:265]: Did you try this test to see whether using playwright magically fixed it?

File: integration/action-test.ts [hunk 9]
```diff
@@ -254,56 +262,54 @@ describe("actions", () => {
...
-  describe("without JavaScript", () => {
-    let restore: Awaited<ReturnType<typeof app.disableJavaScript>>;
-    beforeEach(async () => {
-      restore = await app.disableJavaScript();
-    });
-    afterEach(async () => {
-      await restore();
-    });
+  test.describe("without JavaScript", () => {
+    test.use({ javaScriptEnabled: false });
```
Comment (kentcdodds) [inline integration/action-test.ts:296]: I *think* this should this be in a `beforeAll` hook.

File: integration/action-test.ts [hunk 10]
```diff
@@ -313,31 +319,36 @@ describe("actions", () => {
       expect(html).toMatch(ACTION_DATA_VALUE + " stuff");
     });
 
-    it("rejects too big of an upload", async () => {
+    // TODO: figure out what the heck is wrong with this test...
+    // "Failed to load resource: the server responded with a status of 500 (Internal Server Error)"
+    test.skip("rejects too big of an upload", async ({ page }) => {
```
Comment (kentcdodds) [inline integration/action-test.ts:324]: I'm guessing this is failing for the same reason the other file upload test is failing?

File: integration/helpers/utils.ts [hunk 11]
```diff
@@ -1,5 +1,5 @@
 import prettier from "prettier";
```
Comment (kentcdodds) [inline integration/helpers/utils.ts:1]: I *think* this file is unused.

File: integration/link-test.ts [hunk 12]
```diff
@@ -452,32 +459,42 @@ describe("route module link export", () => {
...
 
-    await restoreJavaScript();
+      await app.goto("/links");
+      await page.waitForSelector('[data-test-id="/links"]');
+      expect(responses.length).toEqual(4);
+    });
   });
 
-  it("preloads assets for other pages and serves from browser cache on navigation", async () => {
+  // playwright doesn't support checking where a resource came from
```
Comment (kentcdodds) [inline integration/link-test.ts:477]: For any of the tests we currently have that playwright won't let us keep, let's delete those.

File: integration/meta-test.ts [hunk 13]
```diff
@@ -1,11 +1,19 @@
...
-describe("meta", () => {
+test.describe("meta", () => {
   let fixture: Fixture;
-  let app: AppFixture;
+  let appFixture: AppFixture;
+
+  // disable JS for all tests in this file
+  // to only disable them for some, add another test.describe()
+  // and move the following line there
+  test.use({ javaScriptEnabled: false });
```
Comment (kentcdodds) [inline integration/meta-test.ts:14]: This should probably be in a `beforeAll` right?

File: integration/playwright.config.ts [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    // {
+    //   name: 'Google Chrome',
+    //   use: {
+    //     channel: 'chrome',
+    //   },
+    // },
+  ],
+
+  /* Folder for test artifacts such as screenshots, videos, traces, etc. */
+  // outputDir: 'test-results/',
```
Comment (kentcdodds) [inline integration/playwright.config.ts:97]: Should we keep this around?

File: integration/playwright.config.ts [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+  /* Run your local dev server before starting the tests */
+  // webServer: {
+  //   command: 'npm run start',
+  //   port: 3000,
+  // },
```
Comment (kentcdodds) [inline integration/playwright.config.ts:103]:
```suggestion
```

File: integration/playwright.config.ts [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+    //   use: {
+    //     channel: 'msedge',
+    //   },
+    // },
+    // {
+    //   name: 'Google Chrome',
+    //   use: {
+    //     channel: 'chrome',
+    //   },
+    // },
```
Comment (kentcdodds) [inline integration/playwright.config.ts:93]:
```suggestion
```

File: integration/playwright.config.ts [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+
+/**
+ * Read environment variables from file.
+ * https://github.com/motdotla/dotenv
+ */
+// require('dotenv').config();
+
+/**
+ * See https://playwright.dev/docs/test-configuration.
+ */
```
Comment (kentcdodds) [inline integration/playwright.config.ts:12]:
```suggestion
```

File: integration/resource-routes-test.ts [hunk 18]
```diff
@@ -49,42 +52,38 @@ describe("loader in an app", () => {
...
-    let restore: Awaited<ReturnType<typeof app.disableJavaScript>>;
-    beforeEach(async () => {
-      restore = await app.disableJavaScript();
-    });
-    afterEach(async () => {
-      await restore();
-    });
-
+  test.describe("without JavaScript", () => {
+    test.use({ javaScriptEnabled: false });
```
Comment (kentcdodds) [inline integration/resource-routes-test.ts:64]: Should this be in `beforeEach`?

**Review by kentcdodds — CHANGES_REQUESTED** [review]
Ran things locally and went through each file. Just a couple things. This is *way* faster 👏👏

**Comment by kentcdodds** [comment]
Oh yeah, could you also update the `contributing.md` file please?

File: integration/action-test.ts (hunk 8 above)
Comment (mcansh) [reply integration/action-test.ts:265]: i did and it didn't 😢

File: .gitignore (hunk 7 above)
Comment (mcansh) [reply .gitignore:13]: it'll only show up if you have `CI=1` it looks like due to https://github.com/remix-run/remix/pull/2634/files#diff-79d1eafd16336460c9185ab9aa820bddafd171a2f65918d8c06968a8e8b69bb4R96-R97, but we probably don't need the trace files, i just let them initially as `npm init playwright` gave it to me

Comment (kentcdodds) [reply .gitignore:13]: Cool. Let's leave it in then 👍

File: integration/action-test.ts (hunk 9 above)
Comment (mcansh) [reply integration/action-test.ts:296]: you'd think, but no 😉 https://playwright.dev/docs/test-configuration#local-configuration

File: integration/action-test.ts (hunk 10 above)
Comment (mcansh) [reply integration/action-test.ts:314]: yupp

File: packages/remix-dev/package.json (hunk 4 above)
Comment (mcansh) [reply packages/remix-dev/package.json]: eh my git fu is failing me today

**Review by kentcdodds — APPROVED** [review]
Solid 👍

**Comment by kentcdodds** [comment]
CI is giving us this: 

image

https://github.com/remix-run/remix/actions/runs/2151038318

**Comment by mcansh** [comment]
> CI is giving us this:
> 
> image
> 
> [remix-run/remix/actions/runs/2151038318](https://github.com/remix-run/remix/actions/runs/2151038318)

forgot to remove the upload action, one sec :)

File: .github/workflows/test.yml [hunk 19]
```diff
@@ -112,10 +112,3 @@ jobs:
 
       - name: 👀 Run Integration Tests
         run: "yarn test:integration"
-
-      - name: 📤 Upload test results
-        if: always()
-        uses: actions/upload-artifact@v3
-        with:
-          name: playwright-results
-          path: test-results
```
Comment (kentcdodds) [inline .github/workflows/test.yml]: What would this do? Would we be able to see the test results somehow? Might be pretty cool to keep if so.

**Review by kentcdodds — APPROVED** [review]
Just one last question, otherwise I think this is ready to go.

File: .github/workflows/test.yml (hunk 19 above)
Comment (mcansh) [reply .github/workflows/test.yml]: on failures, you can have it upload a trace (haven't really looked into that though) and any snapshots that were generated during the test run https://github.com/remix-run/remix/actions/runs/2111669650 look at the bottom for a download link

Comment (kentcdodds) [reply .github/workflows/test.yml]: Nice! I think we'll keep off it for now, but maybe we can bring it in later. Thanks!

**Pull request merged**